### PR TITLE
Desktop platform jail signal callback support

### PR DESCRIPTION
**Pull request by maxhora**
Native library build adjusted to be used by StatusIm Desktop app (by react-native-desktop's native module "Status").

Introduced possibility to set specific C callback method to get new geth jail signals notifications.  

Important changes:
`SetJailSignalCallback` is called by external CPP source code. Preset callback is called from `signal/signals.c` on new jail signal propagation.

Closes https://github.com/status-im/status-react/issues/4841

### CODE REVIEW

File: signal/signals.c [hunk 1]
```diff
@@ -198,18 +198,29 @@ bool StatusServiceSignalEvent(const char *jsonEvent) {
...
+	if (gCallback) {
+		gCallback(jsonEvent);
+	} else {
+		NotifyNode((char *)jsonEvent); // re-send notification back to status node
+	}
 
 	return true;
 }
 
+void SetJailSignalCb(void* cb) {
```
Comment (adambabik) [inline signal/signals.c]: `s/void* cb/void *cb` to keep consistency.

Also, this is not related to jail. Signals are sent from different parts of the app. `SetSignalEventCallback(void *cb)` sounds more appropriate.

**Review by adambabik — COMMENTED** [review]
LGTM but please improve naming. Signals are not only related to jail.

File: signal/signals.go [hunk 2]
```diff
@@ -82,3 +85,8 @@ func NotifyNode(jsonEvent *C.char) {
 func TriggerTestSignal() {
 	C.StatusServiceSignalEvent(C.CString(`{"answer": 42}`))
 }
+
+// SetSignalEventCallback set callback
+func SetSignalEventCallback(cb unsafe.Pointer) {
```
Comment (maxhora) [inline signal/signals.go:90]: @adambabik thank you for the correction, have updated functions naming.

```
s/void* cb/void *cb to keep consistency.

Also, this is not related to jail. Signals are sent from different parts of the app. SetSignalEventCallback(void *cb) sounds more appropriate.
```

Could you please elaborate regarding `s/void* cb/void *cb to keep consistency.` Not sure that I'm getting this correctly.

Also the question regarding naming best practice:
Go function `SetSignalEventCallback` finally calls C function `SetEventCallback` defined in `signals.c`. I would like to name C function identically to function defined in Go (since the purpose of both is the same), but compilation fails if I do that way. I'm wondering if exists some naming convention to point that function is exported from C in title of that function.

Thanks

**Comment by divan** [comment]
@MaxRis thanks for PR. Can you please elaborate more on why this callback should be external?

**Comment by maxhora** [comment]
Hey @divan , since C/C++ seems doesn't provide language build-in tools to reference required custom meta types and set callback directly in signals.c ( like it's done for Android and iOS ), callback in API looks as the simplest solution to achieve this functionality ( currently it's used as following https://github.com/status-im/status-react/blob/desktop/modules/react-native-status/desktop/rctstatus.cpp#L43 )

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/status-im/status-react">status-im/status-react</a></strong></div><div>status-react - a free (libre) open source, mobile OS for Ethereum</div></blockquote>

File: signal/signals.go (hunk 2 above)
Comment (adambabik) [reply signal/signals.go:90]: > `s/void* cb/void *cb` to keep consistency

It's just cosmetics. The star is put next to the argument name in this file so `void SetEventCallback(void *cb)` :)

I don't think we use any convention for that. I assume the compilation problem is because Go code is compiled as CGO and it basically shares the same symbol table. `void SetEventCallback(void *cb)` looks good to me though. I would keep it.

**Review by adambabik — APPROVED** [review]

**Review by adriacidre — APPROVED** [review]
LGTM

File: signal/signals.go (hunk 2 above)
Comment (maxhora) [reply signal/signals.go:90]: @adambabik got it, thanks for clarification! Sorry, haven't understood `s/` correctly
